### ISSUE TITLE
f16 matmul always accumulates in f16, even where an f32-accumulating kernel already exists

### ISSUE THREAD

**Issue by czoli1976** [sonos/tract#2588]
For an f16 model, `strategize` never considers the f32-accumulating kernels tract already has, so every f16 matmul accumulates in f16 over the whole K dimension. At LLM K that is a measurable and compounding accuracy cost, and it is currently decided implicitly rather than as a trade anyone chose.

### What it costs

Median relative error of a dot product against an f64 reference, 200 trials of `a ~ N(0, 1/sqrt(K))`, `b ~ N(0,1)` stored as f16:

| K | f16 accumulate | f32 accumulate |
|---|---|---|
| 256 | 0.17% | 0.0000% |
| 512 | 0.24% | 0.0000% |
| 1024 | 0.41% | 0.0000% |
| 4096 | **0.69%** | 0.0001% |

It grows as roughly sqrt(K) — the partial sum random-walks to about sqrt(K) times a single term, so each add rounds at that scale and the errors accumulate. Every matmul in every layer pays it.

For scale: the accurate-exp softmax in #2318 was reverted for shifting f16 logits by roughly 1e-7 relative. This is four orders of magnitude larger, on every matmul.

### Why it is not simply a missing kernel

The pieces are already there:

- `EinSum::acceptable_accumulators()` (`core/src/ops/einsum/mod.rs:127`) already returns **both** `f16` and `f32` when `operating_dt` is f16, so the design anticipates an f32 accumulator on f16 operands.
- `arm64simd_mmm_f32_32x1_gen<f32>` and `arm64simd_mmm_f32_32x3_gen<f32>` (`linalg/src/arm64/arm64simd.rs:70`, `:80`) each declare `packing[3] = f16f16` and `store(f16)` — f16 in, **f32 accumulate**, f16 out, at `ManuallyOptimized` quality.

What blocks them is the fast path at the top of `strategize` (`core/src/ops/einsum/kernel_selection.rs:24-38`): when the operand dtypes match `operating_dt` and `ops().mmm(operating_dt, m, k, n)` returns a `ManuallyOptimized` kernel, it returns immediately. `Ops::mmm` keys on the accumulator (`linalg/src/lib.rs:206`), so for f16 that is always `mmm_f16`/`mmv_f16` — the `arm64fp16` kernels, which accumulate in f16. The f32-accumulating candidates are only reachable through `list_impls`, which that early return skips, and both sets are `ManuallyOptimized` so the later scoring would not separate them on quality either.

### How to fix

The cheapest concrete step needs no new assembly. At **n = 1** — LLM decode — an f32-accumulating kernel already covers the shape: `arm64simd_mmm_f32_32x1_gen` satisfies every condition `list_impls` checks (internal type `f32` is in `acceptable_accumulators()`, `stores()` contains f16, and `packing[3]` is `f16f16`), yet `mmv_f16` -> `arm64fp16_mmm_f16_128x1_gen` wins by short-circuit. Letting f16 einsums fall through to `list_impls` instead of taking the fast path would make it selectable:

```rust
// core/src/ops/einsum/kernel_selection.rs, in the fast-path guard
&& op.op.operating_dt != f16::datum_type()
```

That alone changes which kernel is picked, so it wants a bench before it lands — `arm64fp16_mmm_f16_128x1_gen` is 128 wide against 32, and the f16 kernels are presumably faster for good reason.

A better shape long term is an explicit policy rather than an accident of selection order: a knob (mirroring the opt-in `TRACT_AMX_BF16`, which trades accuracy *away*) that restricts `acceptable_accumulators()` to `f32` for f16 einsums, so accuracy-sensitive deployments can opt in and the default stays fast. Wide f16-in/f32-accumulate kernels for nr=4/8 would then be a follow-up, only if the knob shows the accuracy is worth the kernel work.

### What I have not established

Whether this converts into end-to-end model quality. That needs the `large_models` RBO nightly, which I cannot run — so the numbers above are per-dot-product, not per-model, and I would not want the fix landed on the strength of them alone. Filing it so the trade is visible and deliberate rather than implicit.

Context: found while auditing low-precision accumulators over long reductions, which also turned up #2586 and #2587.

🍍

**Comment by czoli1976** [comment]
Measured this properly rather than reading it off the code, and **the fix I proposed above does not work**. Corrections and data below.

### Correction 1 — wrong kernels named

On Apple silicon `apple_amx::plug` runs last, so an f16 matmul selects `apple_amx_mmm_f16_64x1` / `apple_amx_mmm_f16_64x32`, not the `arm64fp16_*` kernels I named. Both are declared `<f16>`, so the accumulator is still f16 and the substance of the report stands — but the kernel names above are wrong for this hardware.

### Correction 2 — the one-line guard is not sufficient

Adding `&& op.op.operating_dt != f16::datum_type()` to the fast-path guard only moves the choice to `arm64fp16_mmm_f16_128x1_gen`, which is **also** an f16 accumulator: both candidates are `ManuallyOptimized`, so scoring ties and the 128-wide mr wins the tie-break over the 32-wide f32 kernel.

Restricting `acceptable_accumulators()` to f32 on its own does nothing either, because the fast path returns before that function is ever consulted.

**Both changes together** do reach the f32-accumulating kernels:

```rust
// core/src/ops/einsum/kernel_selection.rs, fast-path guard
&& op.op.operating_dt != f16::datum_type()

// core/src/ops/einsum/mod.rs, acceptable_accumulators()
} else if self.operating_dt == f16::datum_type() {
    tvec!(f32::datum_type())
```

With both, selection becomes `arm64simd_mmm_f32_32x1_gen` at n=1 and `arm64simd_mmm_f32_32x3_gen` at wider n — f16 in, f32 accumulate, f16 out, as claimed.

### What it actually buys and costs

k=4096, error against an f64 reference over the same f16 inputs, and wall time per run:

| shape | f16 acc mean / max | f32 acc mean / max | f16 time | f32 time |
|---|---|---|---|---|
| m=1 n=1 | 2.13% / 2.13% | **0.032%** / 0.031% | 12.1 us | **10.2 us** |
| m=32 n=1 | 2.04% / 24.9% | **0.020%** / 0.041% | 28.4 us | **28.2 us** |
| m=32 n=32 | 3.45% / 781% | **0.018%** / 0.126% | 28.9 us | 143 us |
| m=128 n=128 | 5.75% / 14630% | **0.018%** / 2.81% | 200 us | 2016 us |

The absolute error levels depend on the operand distribution (this probe uses structured rather than random data, and the extreme max errors are near-cancellation cases where the true value is tiny). The robust signals are the **ratio** — roughly two orders of magnitude — and the timings.

### The actionable part

**At n = 1 the f32 accumulator is free or slightly faster and about 100x more accurate.** That is the LLM decode path, and on this hardware there is no trade to make: `arm64simd_mmm_f32_32x1_gen` is not slower than `apple_amx_mmm_f16_64x1`.

At wider n it costs 5-10x, because the only f32-accumulating option is nr=3 against AMX's 64x32. So the policy wants to be shape-aware — f32 accumulation where an equally fast kernel exists, f16 elsewhere — rather than a global switch. Wide f16-in/f32-accumulate kernels would be the follow-up that makes it unconditional, and only then is it worth the assembly.

Still not established: whether this converts end to end. That needs the `large_models` RBO nightly, which I cannot run.

🍍
